### PR TITLE
mat: generalise basic arithmetic vector operations

### PR DESCRIPTION
**Pull request by kortschak**
Please take a look.

Updates #167.

### CODE REVIEW

**Comment by kortschak** [comment]
`MulVec` is more difficult to do in a sensible way.

File: mat/vector.go [hunk 1]
```diff
@@ -503,105 +503,118 @@ func (v *VecDense) DivElemVec(a, b Vector) {
...
 	case *VecDense:
-		if v != a {
-			v.checkOverlap(a.mat)
+		if v != aU {
+			v.checkOverlap(aU.mat)
 		}
 
-		if a.Len() == 1 {
+		if aU.Len() == 1 {
 			// {1,1} x {1,n}
```
Comment (vladimir-ch) [inline mat/vector.go]: `{1,1} x {n,1}`?

File: mat/vector.go [hunk 2]
```diff
@@ -503,105 +503,118 @@ func (v *VecDense) DivElemVec(a, b Vector) {
...
+				return
+			}
 			for i := 0; i < b.Len(); i++ {
-				v.mat.Data[i*v.mat.Inc] = av * b.mat.Data[i*b.mat.Inc]
+				v.mat.Data[i*v.mat.Inc] = av * b.AtVec(i)
 			}
 			return
 		}
 		if b.Len() == 1 {
 			// {1,n} x {1,1}
```
Comment (vladimir-ch) [inline mat/vector.go]: `{n,1} x {1,1}`?

File: mat/vector.go [hunk 3]
```diff
@@ -503,105 +503,118 @@ func (v *VecDense) DivElemVec(a, b Vector) {
...
-			bv := b.At(0, 0)
-			for i := 0; i < a.Len(); i++ {
-				v.mat.Data[i*v.mat.Inc] = bv * a.mat.Data[i*a.mat.Inc]
+			bv := b.AtVec(0)
+			for i := 0; i < aU.Len(); i++ {
+				v.mat.Data[i*v.mat.Inc] = bv * aU.mat.Data[i*aU.mat.Inc]
 			}
 			return
 		}
 		// {n,1} x {1,n}
```
Comment (vladimir-ch) [inline mat/vector.go]: `{1,n} x {n,1}`?

File: mat/vector.go [hunk 4]
```diff
@@ -97,11 +97,11 @@ func NewVecDense(n int, data []float64) *VecDense {
 	}
 }
 
-// SliceVec returns a new VecDense that shares backing data with the receiver.
+// SliceVec returns a new Vector that shares backing data with the receiver.
 // The returned matrix starts at i of the receiver and extends k-i elements.
 // SliceVec panics with ErrIndexOutOfRange if the slice is outside the capacity
 // of the receiver.
-func (v *VecDense) SliceVec(i, k int) *VecDense {
+func (v *VecDense) SliceVec(i, k int) Vector {
```
Comment (vladimir-ch) [inline mat/vector.go]: Why this change?

Comment (kortschak) [reply mat/vector.go]: Interface satisifaction; if there are multiple `SliceVec` implementers they can only satisfy an interface if there is an equal signature. This is what we did with `Slice` on `Dense`. Maybe it needs rediscussion?

File: mat/vector.go (hunk 1 above)
Comment (kortschak) [reply mat/vector.go]: Interesting. If columns of A is 1, rows of b must also be 1. Now we specify that `VecDense` is a column vectore, which means that here we cannot have `{1,1} x {1,n}` unless n is 1.

/cc @btracey

File: mat/vector.go (hunk 2 above)
Comment (kortschak) [reply mat/vector.go]: Fixed

File: mat/vector.go (hunk 3 above)
Comment (kortschak) [reply mat/vector.go]: Fixed

File: mat/vector.go (hunk 1 above)
Comment (kortschak) [reply mat/vector.go]: I'm tightening this up, and as a second commit removing this case since it doesn't improve much over the fall through at the bottom.

File: mat/vector.go [hunk 5]
```diff
@@ -242,247 +261,362 @@ func (v *VecDense) AddScaledVec(a *VecDense, alpha float64, b *VecDense) {
...
+			amat := aU.RawMatrix()
+			// We don't know that a is a *Dense, so make
+			// a temporary Dense to check overlap.
+			(&Dense{mat: amat}).checkOverlap(v.asGeneral())
+			t := blas.NoTrans
+			if trans {
+				t = blas.Trans
+			}
+			blas64.Gemv(t, 1, amat, bmat, 0, v.mat)
+			return
```
Comment (btracey) [inline mat/vector.go]: This needs a fallthrough, or other code for the non fast path.

File: mat/vector.go [hunk 6]
```diff
@@ -242,247 +261,362 @@ func (v *VecDense) AddScaledVec(a *VecDense, alpha float64, b *VecDense) {
...
+		}
+	} else {
+		fast = false
 	}
 
 	v.reuseAs(ar)
 
 	switch {
 	case alpha == 0: // v <- a
+		if v == a {
```
[the remark this answers is not in the document]
Comment (kortschak) [reply mat/vector.go]: It has to come after the `reuseAs` call.

File: mat/vector.go (hunk 5 above)
Comment (kortschak) [reply mat/vector.go]: Like the one after the switch?

Comment (btracey) [reply mat/vector.go]: Oh, sorry.
